docs: guide on scaling your API (#4414)
graphql/graphql-js: website/pages/docs/scaling-graphql.mdx

The following example defines a minimal schema that serves a single `hello` field:

```js
import { GraphQLSchema, GraphQLObjectType, GraphQLString } from 'graphql';

const QueryType = new GraphQLObjectType({
  name: 'Query',
  fields: {
    hello: {
      type: GraphQLString,
      resolve: () => 'Hello from a monolithic schema!',
    },
  },
});

export const schema = new GraphQLSchema({ query: QueryType });
```

This structure works well for small to medium projects, especially when a single team owns the entire
graph. It's simple to test, deploy, and reason about. As long as the schema remains manageable
in size and scope, there's often no need to introduce additional architectural complexity.

## Schema stitching

As your application evolves, you may want to split your schema across modules or services while
still presenting a unified graph to clients. Schema stitching allows you to do this by merging
multiple schemas into one executable schema at runtime.

GraphQL.js does not include stitching capabilities directly, but the
[`@graphql-tools/stitch`](https://the-guild.dev/graphql/stitching/docs/approaches) package
implements stitching features on top of GraphQL.js primitives.

> The following example stitches two subschemas into a single gateway schema:

Reviewer: Let's avoid the term "gateway" here, it implies complexity that need not exist:

```suggestion
The following example stitches two subschemas into a single stitched schema:
```

```js
import { stitchSchemas } from '@graphql-tools/stitch';

export const schema = stitchSchemas({
  subschemas: [
    { schema: userSchema },
    { schema: productSchema },
  ],
});
```

> Each subschema can be developed and deployed independently. The gateway handles query delegation,

Reviewer: ```suggestion
Each subschema can be developed and deployed independently. The stitched schema handles query delegation,
```

merging, and resolution across them.

Stitching is useful when:

- Integrating existing GraphQL services behind a single endpoint
- Incrementally breaking up a monolithic schema
- Creating internal-only aggregators

However, stitching can add runtime complexity and often requires manual conflict resolution for
overlapping types or fields.

## Federation

> Federation is a distributed architecture that composes a single graph from multiple independently

Reviewer: ```suggestion
Federation is a distributed architecture that composes a single GraphQL schema from multiple independently
```

Unlike schema stitching, federation is designed for large organizations where teams need autonomy over
their part of the schema and services must be deployed independently.

Federation introduces a set of conventions to coordinate between services. For example:

- `@key` declares how an entity is identified across subgraphs
- `@external`, `@requires`, and `@provides` describe field-level dependencies across service boundaries

Rather than merging schemas at runtime, federation uses a composition step to build the final schema.
A dedicated gateway routes queries to subgraphs and resolves shared entities.

GraphQL.js does not provide built-in support for federation. To implement a federated subgraph using
GraphQL.js, you'll need to:

- Add custom directives to the schema
- Implement resolvers for reference types
- Output a schema that conforms to a federation-compliant format

Most federation tooling today is based on 
[Apollo Federation](https://www.apollographql.com/docs/graphos/schema-design/federated-schemas/federation). 
However, other approaches exist:

- [GraphQL Mesh](https://the-guild.dev/graphql/mesh) allows federation-like composition across 
services using plugins.

> - Custom gateways and tooling can be implemented using GraphQL.js or other frameworks.

Reviewer: ```suggestion
- Custom gateways and tooling can be implemented using GraphQL.js or other frameworks.
- The [GraphQL Composite Schemas WG](https://github.com/graphql/composite-schemas-wg/) (formed of Apollo, ChilliCream, The Guild, Netflix, Graphile and many more) are working on an open specification for the next generation of GraphQL Federation
```

Federation is most useful when schema ownership is distributed and teams need to evolve their subgraphs
independently under a shared contract.

## Choosing the right architecture

The best structure for your GraphQL API depends on your team size, deployment model, and how your
schema is expected to grow.

| Pattern | Best for | GraphQL.js support | Tooling required |
|---|---|---|---|
| Monolith | Simpler apps and single-team ownership | Built-in | None |
| Schema stitching | Aggregating services you control | External tooling required | `@graphql-tools/stitch`

> | Federation | Distributed graphs across many teams | Manual implementation | Apollo Federation or custom |

Reviewer: I'm concerned that people may think "my app isn't simple, so it needs federation" which can lead to beginners being turned away from GraphQL due to jumping too quickly into complexity. As you mention elsewhere, Federation is really there to help large enterprises with lots of teams working independently of each other, but all contributing to the same schema; I've attempted to capture this:

```suggestion
| Pattern | Best for | GraphQL.js support | Tooling required |
|---|---|---|---|
| Monolith | Default choice for most projects; simpler, faster, easier to reason about | Built-in | None |
| Schema stitching | Aggregating services you control | External tooling required | `@graphql-tools/stitch`
| Federation | Large enterprises; many teams contributing to a distributed graph independently | Manual implementation | Significant tooling and infrastructure |
```
